README
This is a "plugin" for the Video Disk Recorder (VDR).

Originally Written by:       herbert attenberger <herbsl at a-land.de>

Now supported by:            daniel meyerholt <dxm523@googlemail.com>

Original Project's homepage: <http://www.a-land.de>

Current homepage:
              <http://projects.vdr-developer.org/projects/plg-vdrrip>

See the file COPYING for license information.

See the file INSTALL for the installation instructions.


Description:
============

Vdrrip is a plugin to encode the vdr-recordings or dvd's into several
formats. The supported output-formats are:

Containers:         avi, ogm, matroska
Video-codecs(divx): lavc, xvid, divx4/5
Audio-codecs:       mp3, ogg-vorbis, stream-copy for ac3 and mpeg2

Ogg-vorbis, ac3, ogm and matroska have to be manually activated in the
setup-menu of the plugin. All further codecs are automatically detected by
the plugin.

You can manually adjust the encoding values, or simply select a
template (read the section 'Templates' in this manual to find out
more about them). If you input the resulting movie size (or number of
files), the video/audio bitrate is calculated. If you change the video/
audio bitrate, the file size is calculated. If there is more then one
audio-stream available, you can toggle between them.

You can choose between four scaling mechanism:

off:     Don't scale, use the original resolution
auto:    The scale values are calculated on a algorithm which is based on
         the bpp-value. Read the file DOCS/tech/encoding-tips.txt in your
	 MPlayer-source directory for further informations. The aspect ratio
	 is corrected
dvb:     You can only choose values, which are supported by the DVB-Card
         without needing a further scaling while playing. This option
	 (together with streamcopy as audio codec) is very useful for slow
         computers (it is possible to play those files on my Celeron 466
         Mhz).
manual:  Select the resolution manually.

In the auto and manual-mode you can automatically detect and crop the black
movie-boarders with the green key. All settings are saved in the file
save.vdrrip in the movie directory. After this you can add the movie with
the red key to the encoding queue (this is the file queue.vdrrip in the
plugins-directory) and select between preview or full encoding-mode.
The script queuehandler reads the data from the queue, joins
or demuxes the vdr-files and encodes the movie.

Some parameters of vdrrip can be adjusted in the setup menu.

DVD's:
======

See the INSTALL-file how to compile vdrrip with DVD-support.

Vdrrip automatically takes the longest title of the DVD. It is possible
to switch between them in the menu. If you press <ok> on the title-
line, a more comfortable title-menu is opened. The last settings
are saved in the file /tmp/<DVD-NAME>.vdrrip, and will be restored
if you want to encode the same dvd again.

WARNING !

Most DVDs are encrypted with CSS (Contents Scrambling System). Copying
of encrypted DVDs is illegal in many countries. This program is not
meant for those who intend on breaking copyright laws. Every illegal
use of this software is strictly prohibited. In no way I will be held
responsible if you do. Be aware to check your countries law !


Templates:
==========

Within a template ('Vdrrip - edit template') you can predefine the encoding-
settings (either with a fixed file size or a fixed bitrate). The default
settings are taken from the template 'default'. You can add further
templates (e.g 1/2-CD, 2/3-CD ...). All templates are saved in the file
templates.vdrrip in the dir <vdr-config-dir>/PLUGINS.


usage of the quehandler-script:
===============================

queuehandler queuefile tempdir

The queuefile is the file queue.vdrrip in the dir <vdr-config-dir>/PLUGINS.
This file is only created, if you add a movie to the queue. After the
encoding the file is deleted by the queuehandler-script. If the movie
couldn't be encoded by the queuehandler because of an error, the queue-
data will be moved to the file queue.vdrrip.rejected. Then you have to look
into the syslog, correct the error and move the movie back to the queue
with the commands: "cat queue.vdrrip.rejected >> queue.vdrrip" and
"rm queue.vdrrip.rejected"

Tempdir is the directory where all temporary files are created and where the
encoded movies are saved. There should be enough space free !

There is no audio-stream in the preview-mode, if the container type is ogm
or matroska !


The file <vdr-config>/PLUGINS/encode.vdrrip is created while an active
encoding process. You can use it to keep vdr from an automatic shutdown.
I added my sample scripts vdrshutdown.sh and sleephalt.sh in the dir
scripts. Feel free to use them. Read the file FAQ to see how you can add
the queuehandler to the file commands.conf.


Please read the files INSTALL, FAQ and TODO for further informations.
And beware of the dangerous xaver.jpg ...


HISTORY
VDR Plugin 'vdrrip' Revision History
------------------------------------

2003-09-14: Version 0.0.7

- Initial revision.

2003-09-15: Version 0.0.7a

- added -frames 0 to get -identify in MPlayer1.0pre1 running
- some changes in the README and FAQ - file

2003-09-15: Version 0.0.7b

- vdr-recordings can only be selected with selectvdrrip.sh,
  if the file marks.vdr is available
- menue changes in menu-vdrrip.c:
  VDR to 'encode vdr-rec. (Movie-Name)' (it is only
  displayed if a recording is selected
- a lot of changes in the error handling in movie.c:
  (vdrrip shouldn't crash, even if mplayer/mencoder isn't
  available)

2003-09-18: Version 0.0.8

- add pclose(p) in cMovieBase::queryCodecs() to avoid zombie-processes
  (reported by Chrillepup@www.vdrportal.de)
- set the movie names in brackets at the #defines in movie.c
  (reported by zuse@www.vdrportal.de)
- (first) code-cleanup of menu-vdrrip.c
- don't save/delete save.vdrrip if the video/audio-codec isn't detected
  (reported by dbox.network@www.vdrportal.de)
- add (optional) renaming of encoded vdr-recordings
  (thx to memed@www.vdrportal.de)
- selectvdrrip.sh works if the video-directory is a subdirectory
  (thx to memed@www.vdrportal.de)
- 001.vdr is copied, if a symbolic-link could not created (for fat-partitions)
  (reported by memed@www.vdrportal.de)
- add (alternative) usage of svdrpsend.pl in queuehandler.sh
- fixed OSD-bugs (help-keys etc.)
- add experts-menu (with postprocess-filters)
- add handling of postprocess-filters to queuehandler.sh

2003-10-12: Version 0.0.9
- add command line arguments for mplayer/mencoder-location, to avoid
  problems with $PATH (see ./vdr --help)
- fix a bug in menu-vdrrip.c (with postprocess-filters)
- fix a bug in menu-vdrrip.c (display Audio-ID, if more then one audio-id is available)
- add templates-handling (templates.c)
- moved codecs-handling to codecs.c
- removed the bitrate-line from osd
- some changes in the README and FAQ - file

2003-11-24: Version 0.1.0
- code-cleanup of codecs.c
- add ScaleType & Bpp to templates-handling
- write all messages and errors to syslog instead of stdout/stderr
- add Bpp-Value to expert-menu
- add internationalization support
- changed " to ' at the #defines in movie.c
  (thx to memed@www.vdrportal.de)
- changed menu-handling (selectvdrrip.sh is now obsolete)
- some changes in the README and FAQ - file

2003-12-14: Version 0.1.1
- add italian translation (thx to tinconn@virgilio.it)
- moved queue-handling to queue.c
- add locking-mechanism to allow the editing of the queue
- changed vdrriprecordings.c to display recordings with more than
  two subdirectories (reported by frissi@www.vdrportal.de)
- some cosmetic changes of the menus in menu-vdrrip.c

2004-01-17: Version 0.2.0
- add ogg-vorbis support
- add ac3-support
- add ogm-support
- add matroska-support
- now display more audio-informations
- display container-type in queuefile-menu
- get movie length from index.vdr instead of marks.vdr
  (the template "default-fixed Bitrate" is now obsolete)
- add a ac3 MPlayer-patch to the /patches directory
- avoid VideoBitrates < 150 in cMovieBase::setBitrate
- avoid AudioBitrates < 32 in cMovieBase::setBitrate
- decrease VideoBitrate if AudioBitrate is increased
- some cosmetic changes of the menus in menu-vdrrip.c
- add french translation (thx to mak_addikt@yahoo.fr)
- splitted the install-section from README to INSTALL
- lot's of changes in the queuehandler.sh

2004-01-20: Version 0.2.0a
- fixed a bug in cMovie::setCodecs (thx to markus.maeder@oase.ch)
- fixed a bug in queuehandler.sh (reported by cosmo@www.vdrportal.de)

2004-04-22: Version 0.3.0
- add dvd as input source
- add one line in i18.c to be compatible with vdr-1.3.2 and newer
- changed cMovie::setCropValues() to support dvd's
- switch between 'crop boarders' and 'reset boarders' possible
- now round crop & scale-values to a multiple of 16 (to avoid errors with
  xvid and divx)
- add more deinterlacing filters (requested by sebastian.kemper@web.de)
- now colored items are used to display informations
- decide between preview or full mode while adding a movie to the queue
- close the encoding-menue after adding the movie to the queue (requested
  by mariom@www.vdrportal.de)
- display the encoding step in the queue menue (requested by
  skyalex@www.vdrportal.de)
- switching between preview mode and normal mode is possible in the
  queue menu
- moved 2-pass logfiles of the queuehandler to $tempdir.
  (requested by mrmoe@www.vdrportal.de)
- check for mplayer&mencoder (requested by hapemo@www.vdrportal.de)
- some cosmetic changes in queue.c

2010-02-06: Version 0.3.2
- Changed maintainer
- Merged many patches from e-tobi's debian source package:
    * patches by Thomas Günther <tom@toms-cafe.de>
      - maketempdir patch to create temporary directory
      - greppid2 patch to detect second running process instead of third
      - fix ogm/ac3 vdrsync 0.1.2.2dev1 creates bd.mpa, not bd.ac3 when
        demuxing
      - preserve owner of queue.vdrrip in queuehandler.sh
      - fix identification of aspect ratio
        (http://vdrportal.de/board/thread.php?threadid=53225 and
         http://vdrportal.de/board/thread.php?threadid=58775)
      - change VDRVERSION define to APIVERSION in Makefile
      - fix some menus for vdr>=1.3.7 (by user norad at vdrportal.de & thomas)
    * patch by Herbert Attenberger <herbsl at a-land.de> and user micmac 
      at vdrportal.de to change dvd-parameter to -dvd if the mencoder-version
      is 0.XX
    * patch by Stefan Wagner to change mencoder parameters when detecting
      cropping
    * patches by Tobias Grimm <etobi@debian.org>
      - use libdvdread's NEW_IFO_READ functions
      - fix compilation for gcc4.4
- Needs Project-X for demux.

2014-06-21: Version 0.5 WIP
- Version bump
- Stole i18n Patch from e-tobi
- Make gcc happy by using (char *) casts and inserting some paranthesis
- Fixed build for vdr > 2.1.2



queue.c
//
// queue.c
//

#include <stdio.h>
#include <stdlib.h>
#include <vdr/plugin.h>

#include "queue.h"
#include "a-tools.h"

cQueue::cQueue() {
  Q = NULL;

  WaitUnlock();
  if (! Locked) {Lock();}
  Load();  
}

cQueue::~cQueue() {
  if (! Locked) {Unlock();}
  DELETE(Q);
}

void cQueue::Load() {
  char *buf = NULL;
  size_t i = 0;
  int c = 0;

  FILE *q = fopen(AddDirectory(cPlugin::ConfigDirectory(), QUEUEFILE), "r");
  if (q) {
    // read QueueData
    while (getline(&buf, &i, q) != -1) {
      // reserve memory for QueueData
      Q = (struct QueueData*)realloc(Q, (c + 1) * sizeof(struct QueueData));
      Q[c].Dir          = strcol(buf, ";", 1);
      Q[c].Name         = strcol(buf, ";", 2);
      Q[c].FileSize     = atoi(strcol(buf, ";", 3));
      Q[c].FileNumbers  = atoi(strcol(buf, ";", 4));
      Q[c].VCodec       = strcol(buf, ";", 5);
      Q[c].BitrateVideo = atoi(strcol(buf, ";", 6));
      Q[c].MinQuant     = atoi(strcol(buf, ";", 7));
      Q[c].MaxQuant     = atoi(strcol(buf, ";", 8));
      Q[c].CropWidth    = atoi(strcol(buf, ";", 9));
      Q[c].CropHeight   = atoi(strcol(buf, ";", 10));
      Q[c].CropPosX     = atoi(strcol(buf, ";", 11));
      Q[c].CropPosY     = atoi(strcol(buf, ";", 12));
      Q[c].ScaleWidth   = atoi(strcol(buf, ";", 13));
      Q[c].ScaleHeight  = atoi(strcol(buf, ";", 14));
      Q[c].ACodec       = strcol(buf, ";", 15);
      Q[c].BitrateAudio = atoi(strcol(buf, ";", 16));
      Q[c].AudioID      = atoi(strcol(buf, ";", 17));
      Q[c].PPValues     = strcol(buf, ";", 18);
      Q[c].Rename       = atoi(strcol(buf, ";", 19));
      // migrate from version 0.1.1
      char *cont        = strcol(buf, ";", 20);
      cont ? Q[c].Container = cont : Q[c].Container = strdup("avi");
      // migrate from version 0.2.0a
      char *prev        = strcol(buf, ";", 21);
      prev ? Q[c].Preview = atoi(prev) : Q[c].Preview = 0;

      FREE(buf);
      c++;
    }

    fclose(q);
  } else {dsyslog("[vdrrip] could not open file %s, the queue is probably empty !", QUEUEFILE);}

  NumMovies = c;
}


bool cQueue::Save() {
  FILE *q;
  int c;
  int n = 0;

  if (Locked) {
    Load();
    return false;
  }

  q = fopen(AddDirectory(cPlugin::ConfigDirectory(), QUEUEFILE), "w");
  if (q) {
    for (c = 0; c < NumMovies; c++) {
      if (strcmp(Q[c].Name, "delete") != 0) {
        fprintf(q,"%s;%s;%i;%i;%s;%i;%i;%i;%i;%i;%i;%i;%i;%i;%s;%i;%i;%s;%i;%s;%i\n",
        Q[c].Dir, Q[c].Name, Q[c].FileSize, Q[c].FileNumbers, Q[c].VCodec,
        Q[c].BitrateVideo, Q[c].MinQuant, Q[c].MaxQuant, Q[c].CropWidth,
        Q[c].CropHeight, Q[c].CropPosX, Q[c].CropPosY, Q[c].ScaleWidth,
        Q[c].ScaleHeight, Q[c].ACodec, Q[c].BitrateAudio, Q[c].AudioID,
        Q[c].PPValues, Q[c].Rename, Q[c].Container, Q[c].Preview);
	n++;
      }
    }
    NumMovies = n;
    
    fclose(q);

    // delete queuefile if it's empty
    if (NumMovies < 1) {
      remove(AddDirectory(cPlugin::ConfigDirectory(), QUEUEFILE));
    }
  } else {dsyslog("[vdrrip] could not save %s", QUEUEFILE);}

  Load();

  return true;
}

bool cQueue::New(struct QueueData *q) {
  if (Locked) {return false;}

  Q = (struct QueueData*)realloc(Q, ((NumMovies + 1) * sizeof(struct QueueData)));
  Q[NumMovies] = *q;

  NumMovies++;
  Save();

  return true;
}


bool cQueue::Del(int i) {
  if (i >= 0 && i < NumMovies) {

    // don't delete the first entry of the queuefile,
    // if there is an aktiv encoding
    if (i == 0 && IsEncoding()) {return false;}

    isyslog("added delete flag on movie %s", Q[i].Name);
    Q[i].Name = strdup("delete");
    Save();

    return true;
  }
  return false;
}

bool cQueue::Up(int i) {
  if (i >= 1 && i < NumMovies) {

    // don't move to the first entry of the queuefile,
    // if there is an aktiv encoding
    if (i == 1 && IsEncoding()) return false;

    struct QueueData q;
    q = Q[i];
    Q[i] = Q[i-1];
    Q[i-1] = q;
    Save();

    return true;
  }
  return false;
}

bool cQueue::Down(int i) {
  if (i >= 0 && i < NumMovies - 1) {

    // don't move the first entry of the queuefile,
    // if there is an aktiv encoding
    if (i == 0 && IsEncoding()) return false;

    struct QueueData q;
    q = Q[i];
    Q[i] = Q[i+1];
    Q[i+1] = q;
    Save();

    return true;
  }
  return false;
}

bool cQueue::Switch(int i) {
  if (i >= 0 && i < NumMovies) {

    // don't switch the first entry of the queuefile,
    // if there is an aktiv encoding
    if (i == 0 && IsEncoding()) return false;

    (Q[i].Preview == 0) ? Q[i].Preview = 1 : Q[i].Preview = 0;
    Save();

    return true;
  }
  return false;
}

    
int cQueue::getNumMovies() {return NumMovies;}

struct QueueData* cQueue::getData(int i) {
  if (i >= 0 && i < NumMovies) {return &Q[i];
  } else {return NULL;}
}

char *cQueue::getName(int i) {
  if (i >= 0 && i < NumMovies) {return Q[i].Name;
  } else {return NULL;}
}

char *cQueue::getShortname(int i) {
  if (i >= 0 && i < NumMovies) {
    if (strlen(Q[i].Name) > 20) {
      char *s, *s1;
      s = strsub(Q[i].Name,1 , 17);
      asprintf(&s1, "%s...", s);
      return s1;
    } else {return Q[i].Name;}
  } else {return NULL;}
}

bool cQueue::getLockStat() {return Locked;}

void cQueue::Lock() {
  FILE *l;
  l = fopen(AddDirectory(cPlugin::ConfigDirectory(), LOCKFILE), "w");
  if (l) {
    fprintf(l,"0");
    isyslog("[vdrrip] queuefile locked");
    fclose(l);
  } else {
    dsyslog("[vdrrip] could not lock queuefile");
  }   
}

void cQueue::Unlock() {
  int r = remove(AddDirectory(cPlugin::ConfigDirectory(), LOCKFILE));
  if (r == -1) {dsyslog("[vdrrip] could not unlock queuefile");
  } else {isyslog("[vdrrip] queuefile unlocked");}
}

void cQueue::WaitUnlock() {
  FILE *l;
  int i = 0;
  bool loop = true;

  while (loop) {
    l = fopen(AddDirectory(cPlugin::ConfigDirectory(), LOCKFILE), "r");
    if (l) {

      // check if the content of the lockfile 0
      int c = fgetc(l);
      if (c == 48) {
	loop   = false;
	Locked = false;
	break;
      }
      
      i++;
      if (i > 2) {
	loop   = false;
	Locked = true;
      }
      isyslog("[vdrrip] %d. try: queuefile is locked by another process", i);
      sleep(1);
      fclose(l);
    } else {
      isyslog("[vdrrip] queuefile is not locked by another process");
      loop   = false;
      Locked = false;
    }
  }
}

bool cQueue::IsEncoding() {
  FILE *e;
  e = fopen(AddDirectory(cPlugin::ConfigDirectory(), ENCODEFILE), "r");
  if (! e) return false;
  
  fclose(e);
  return true;
}

char *cQueue::getQueueStat() {
  char *buf = NULL;
  size_t i = 0;

  FILE *e = fopen(AddDirectory(cPlugin::ConfigDirectory(), ENCODEFILE), "r");
  if (e) {
    if (getline(&buf, &i, e) != -1) {
      fclose(e);
      return buf;
    }
  }
  return NULL;
}
